programs/bpf_loader/src/lib.rs

```diff
@@ -1337,6 +1340,162 @@ fn process_loader_upgradeable_instruction(
+                        ic_logger_msg!(log_collector, "Upgrade authority did not sign");
+                        return Err(InstructionError::MissingRequiredSignature);
+                    }
+                    programdata
+                        .get_data()
+                        .len()
+                        .saturating_sub(UpgradeableLoaderState::size_of_programdata(0))
+                }
+                _ => {
+                    ic_logger_msg!(log_collector, "Invalid ProgramData account");
+                    return Err(InstructionError::InvalidAccountData);
+                }
+            };
+            let programdata_key = *programdata.get_key();
+            let programdata_funds = programdata.get_lamports();
+            drop(programdata);
+
+            // Verify Program account
+            let mut program =
+                instruction_context.try_borrow_instruction_account(transaction_context, 1)?;
+            #[allow(deprecated)]
+            if !invoke_context
+                .get_feature_set()
+                .is_active(&remove_accounts_executable_flag_checks::id())
+                && !program.is_executable()
+            {
+                ic_logger_msg!(log_collector, "Program account not executable");
+                return Err(InstructionError::AccountNotExecutable);
+            }
+            if !program.is_writable() {
+                ic_logger_msg!(log_collector, "Program account not writeable");
+                return Err(InstructionError::InvalidArgument);
+            }
+            if program.get_owner() != program_id {
+                ic_logger_msg!(log_collector, "Program account not owned by loader");
+                return Err(InstructionError::IncorrectProgramId);
+            }
+            if let UpgradeableLoaderState::Program {
+                programdata_address,
+            } = program.get_state()?
+            {
+                if programdata_address != programdata_key {
+                    ic_logger_msg!(log_collector, "Program and ProgramData account mismatch");
+                    return Err(InstructionError::InvalidArgument);
+                }
+            } else {
+                ic_logger_msg!(log_collector, "Invalid Program account");
+                return Err(InstructionError::InvalidAccountData);
+            }
+            program.set_data_from_slice(&[])?;
+            let program_key = *program.get_key();
+            drop(program);
+
+            invoke_context.native_invoke(
+                system_instruction::transfer(&programdata_key, &program_key, programdata_funds)
+                    .into(),
+                &[],
+            )?;
+
+            invoke_context.native_invoke(
+                system_instruction::assign(&program_key, &loader_v4::id()).into(),
+                &[],
+            )?;
+
+            invoke_context.native_invoke(
+                solana_loader_v4_interface::instruction::truncate_uninitialized(
```

Comment:
Does truncate also increase the account size? If so, the name might be misleading.

Reply:
Yes, `Truncate` can also increase the size and it is common for file systems to only have one function for both directions: https://www.man7.org/linux/man-pages/man2/truncate.2.html

Reply:
I see linux use the same naming, but we can improve our designation and call it `resize`.
Truncate means shortening: https://www.collinsdictionary.com/dictionary/english/truncate